wandb/proto/wandb_internal.proto

```diff
@@ -424,6 +424,11 @@ message ArtifactManifestEntry {
   string birth_artifact_id = 7;
   // Whether to avoid copying/moving files to the cache while uploading.
   bool skip_cache = 8;
+  enum Policy {
+    MUTABLE = 0;
+    IMMUTABLE = 1;
+  }
+  Policy policy = 9;
```

Comment:
this could use documentation

on the policy field: generally speaking, what's the purpose of this?

on each enum value: what's the effect of this particular policy?

Reply:
Where would you like me to add documentation? I figured there was sufficient documentation around this in `wandb/sdk/artifacts/artifact.py`?

Reply:
In general, this is what the current state of things are and what we'd like to achieve:

Current behavior:

- upload: make a staging copy synchronously ([code](https://github.com/wandb/wandb/blob/08d5cd26ab8a37b8ce2f87be405fef122493cea8/wandb/sdk/artifacts/artifact.py#L1468)), then asynchronously upload and copy to cache ([code](https://github.com/wandb/wandb/blob/08d5cd26ab8a37b8ce2f87be405fef122493cea8/wandb/sdk/artifacts/storage_policies/wandb_storage_policy.py#L385)) and delete the staging copy ([code](https://github.com/wandb/wandb/blob/08d5cd26ab8a37b8ce2f87be405fef122493cea8/wandb/sdk/artifacts/artifact_saver.py#L289))
- download: write to cache ([code](https://github.com/wandb/wandb/blob/08d5cd26ab8a37b8ce2f87be405fef122493cea8/wandb/sdk/artifacts/storage_policies/wandb_storage_policy.py#L141)), copy to user directory ([code](https://github.com/wandb/wandb/blob/08d5cd26ab8a37b8ce2f87be405fef122493cea8/wandb/sdk/artifacts/artifact_manifest_entry.py#L136))

Use cases:

- user writes model checkpoints, they want training to keep going, and the checkpoint to be uploaded and deleted, with no cache or staging
- user has a large dataset which won't change, they want it to be uploaded and left as is, with no cache or staging

Proposal:

Flags for upload path:

- skip_cache: bool = True: whether to cache
- policy: "mutable" | "immutable" | "delete" = "mutable":
    - "mutable": copy to staging synchronously and
        - if skip_cache=False: move from staging to cache after uploading
        - if skip_cache=True: delete from staging after uploading
    - "immutable": require the file to not change and
        - if skip_cache=False: link from cache synchronously
        - if skip_cache=True: nothing
    - "delete": move to staging synchronously and
        - if skip_cache=False: move from staging to cache after uploading
        - if skip_cache=True: delete from staging after uploading


The default is as before (but more efficient: we make one copy instead of two). For the use cases above, the user would set:
model checkpoints: cache=False, policy="delete"
dataset: policy="immutable"

Reply:
I added comments to the proto file...